File: Lab02/nim.py
# These are the functions that take care of the calculations and mechanics of the NIM game
# @Author Ashwin Kalyan

def create_game_state(size, token_max):
    """
    This function initiates the game_state and stores all the relevant data for the game in a list, including the number of tokens per row. 

    Args:
        size (int): How many rows (elements) the game (game_state list) will have.
        token_max (int): The maximum amount of tokens per row.

    Return:
        list: A list of how many tokens are in each row, where a row corresponds to an index in the list.
    """
    
    global game_state 
    game_state = []

    adjusted_size = size + 1 # enables range() to properly count from 1 to size # without exclusion issues

    if size == 0:
        return []
    
    for i in range(1, adjusted_size):
        if i <= token_max: # checks if the position value has reached the max token value
            game_state.append(i) # if not, we want to append the position value
        elif i >= token_max:
            game_state.append(token_max) # if so, we want to append the max token value
    
    return game_state

def is_valid_move(game_state, row, takes):
    """
    This function makes sure a user's move does not violate any rules, including being out of bounds and taking a maximum of 3 tokens at a time.

    Args:
        game_state (list): A list of how many tokens are in each row, where a row corresponds to an index in the list.
        row (str): Which row the user wants to take from.
        takes (str): How many tokens the user wants to take.

    Return:
        bool: A boolean that determines whether the move is legal or not: True for it is, False for it is not.
    """

    if not row.isdigit() or not takes.isdigit(): # checks if the strings row and takes are valid digits
        return False
    
    row = int(row)
    takes = int(takes)

    if row < 1 or row > len(game_state): # check if the row is within the valid range (doesn't exceed the number of rows in game_state)
        return False
    
    if takes < 1 or takes > 3: # check if the number of tokens to take is between 1 and 3 per game rules
        return False

    if takes > game_state[row - 1]: # check if the number of tokens to take is not greater than the tokens in the row
        return False 

    return True # if all checks pass, the move is valid

def update(game_state, row, takes):
    """
    This function creates a copy of game_state and updates the values within it according to the user's move.

    Args:
        game_state (list): A list of how many tokens are in each row, where a row corresponds to an index in the list.
        row (int): Which row number needs to be updated.
        takes (int): How many tokens to take away.

    Return:
        list: An updated list of how many tokens are in each row, where a row corresponds to an index in the list.
    """
    
    updated_game_state = list(game_state) # create a new copy of the game_state list
    updated_game_state[row] = updated_game_state[row] - takes # in the copy, update the specified row by subtracting the current value by the takes value

    return updated_game_state

def draw_game_state(game_state):
    """
    Draws the game state in the form of the row number and its associated number of tokens.

    Args:
        game_state (list): A list of how many tokens are in each row, where a row corresponds to an index in the list.

    Return:
        None
    """
    print("=" * 20) # starting border

    for row in range(len(game_state)): # loops through each row in game_state
        print("{} ".format(row + 1) + "#" * game_state[row]) # this prints the row number adjusted by one (since we don't want a row 0), then prints the # needed according to the value stored at row in game_state

    print("=" * 20) # ending border

def is_over(game_state):
    """
    Checks if the game has ended by determining whether all rows in game_state are empty or not.

    Args:
        game_state (list): A list of how many tokens are in each row, where a row corresponds to an index in the list.

    Return:
        bool: A boolean that determines if a game is over or not.
    """
    for row in range(len(game_state)):
        if game_state[row] != 0: # checks if row is not empty
            return False
    
    return True

File: Lab06/NIM.java
/*
 * CSCI 1913
 * Lab 06 
 * Author: Ashwin Kalyan
 * */


public class NIM {
    /**
     * cerate a game state array
     * @param size -- the number of rows
     * @param stickMax -- the largest value
     * @return an array representing a token array. The array in the first position will be one, each following will be
     *     one larger up to the max.
     */
    public static int[] createGameState(int size, int stickMax){
        int[] game_state = new int[size];
        
        if (size == 0) {
            return new int[0]; // return an empty array if size is 0
        }

        for (int i = 0; i < size; i++) {
            if (i + 1 <= stickMax) { 
                game_state[i] = i + 1; // if the position hasn't reached the max token value, set to i
            } else if (i >= stickMax) {
                game_state[i] = stickMax; // if so, set the value to max token value 
            }
       }

        return game_state;
    }

    /**
     * This provided function operators similarly to the python isDigit method. You give it a string and it will tell
     * you if the string contains only digits. Give it a read -- the actual design isn't hard, basically a linear search.
     * @param str any string
     * @return true if all letters in the string are digits.
     */
    private static boolean isDigit(String str) {
        for (int i = 0; i < str.length(); i++) {
            if (!Character.isDigit(str.charAt(i))){
                return false;
            }
        }
        return true;
    }

    /**
     * Checks if a given move (as represented by two input strings) is valid. This should check if the inputs are numbers
     * if those numbers are in the valid range.
     * @param gameState an array representing the number of tokens in each row.
     * @param row a string representing which row the user wants to take from. 1-indexed.
     * @param takes a string representing how many tokens the user wants to take.
     * @return true if and only if the move would be valid
     */
    public static boolean isValidMove(int[] gameState, String row, String takes) {
        if (!isDigit(row) || !isDigit(takes)) {
            return false;
        }

        int row_int = Integer.parseInt(row);
        int takes_int = Integer.parseInt(takes);

        if (row_int < 1 || row_int > gameState.length) { // Check if the row is within the valid range
            return false;
        }

        if (takes_int < 1 || takes_int > 3) { // Check if the number of tokens to take is between 1 and 3
            return false;
        }

        if (takes_int > gameState[row_int - 1]) { // Check if the number of tokens to take is not greater than the tokens in the row
            return false;
        }

        return true; // If all checks pass, the move is valid
    }

    /**
     * User System.out.println to represent a token grid to the program user.
     * @param gameState an array representing the number of tokens in each row.
     */
    public static void drawGameState(int[] gameState){
        System.out.println("====================");

        for (int i = 0; i < gameState.length; i++) {
            System.out.print(i + 1 + " "); // Print the row number
            for (int j = 0; j < gameState[i]; j++) {
                System.out.print("#"); // Print the number of tokens in the row
            }
            System.out.println();
        }

        System.out.println("====================");
    }


    /**
     * Create an updated version of the game state. You can assume the row and takes are valid for the gameState array provided.
     * @param gameState an array representing the number of tokens in each row.
     * @param row the row the user wants to take from (0-indexed)
     * @param takes the number of tokens the user wants to take
     * @return a new array representing the state number of tokens in each row after the given numbers were removed.
     */
    public static int[] update(int[] gameState, int row, int takes){
        int[] updatedGameState = new int[gameState.length]; // Create a new array with the same length as gameState

        // Copy the contents of gameState into updatedGameState
        for (int i = 0; i < gameState.length; i++) {
            updatedGameState[i] = gameState[i];
        }

        // Update the specified row by subtracting the takes value
        updatedGameState[row] = updatedGameState[row] - takes;

        return updatedGameState;
    }

    /**
     *
     * @param gameState an array representing the number of tokens in each row.
     * @return true if and only if every element of gameState is false.
     */
    public static boolean isOver(int[] gameState) {
        for (int i = 0; i < gameState.length; i++) {
            if (gameState[i] != 0) {
                return false; // If any row has tokens, the game is not over
            }
        }

        return true; // If all rows are empty, the game is over
    }
}